tests/cpp/common/test_io.cc

```diff
@@ -39,5 +44,50 @@ TEST(IO, FixedSizeStream) {
     ASSERT_EQ(huge_buffer, out_buffer);
   }
 }
+
+
+#if SIZE_MAX == 0xFFFFFFFFFFFFFFFF  // Only run this test on 64-bit system
+TEST(IO, LoadSequentialFile) {
```

Comment:
Could you please clarify the bug a bit more?  I'm confused.

> The value returned may be less than nbyte if the number of bytes left in the file is less than nbyte

Can we rephrase it to: maximum returned value is actual bytes left in the file being read?  If yes, then what's the condition to trigger the bug that we are trying to fix?  Since as long as we don't try to read extra non-existing bytes, the returned value should be correct no matter how large is the file right?  And here we estimate the file size by:
```
  size_t f_size_bytes = fs.st_size;
  buffer.resize(f_size_bytes + 1);
```
So, if `fs.st_size` is a correct value, then we are reading exactly the number of bytes in file. Why would we end up with an half loaded buffer?


Looking at the test, it seems the real cause is (quoting from the link you provided):

> If the value of nbyte is greater than {SSIZE_MAX}, the result is implementation-defined.

Can we add an optional parameter to the load function specifying batch size for test, instead of creating a huge file?

Reply:
Also, as now we have replaced the number printing algorithm, generated file should be much much smaller.

Reply:
The bug has nothing to do with SSIZE_MAX. The bug has to do with the behavior of read(). When we call read() function with nbytes=2.3GB, it does not actually read 2.3GB. So we must call read() twice or more to get all the bytes into the buffer; otherwise we end up with half full buffer.

Reply:
> Can we add an optional parameter to the load function specifying batch size for test, instead of creating a huge file?

Not sure what you mean here. How can we test the behavior of read() with nbytes=2.3GB without creating a tempfile of size 2.3GB?

Reply:
> Can we rephrase it to: maximum returned value is actual bytes left in the file being read?

No. The standard lists a few possible scenarios where read() returns fewer bytes than the number of bytes requested (nbytes parameter). One possible scenario is to have a short file. For this particular bug, we do have a complete file, yet read() returns fewer bytes. I honestly don’t understand why read() gives fewer than 2.3GB.

This behavior is quite well known. The prevailing wisdom in UNIX community is to wrap read() in a while loop.

Reply:
> Not sure what you mean here. How can we test the behavior of read() with nbytes=2.69 GB without creating a tempfile of size 2.69 GB?

What I'm worrying about is the behaviour is dependent on other factors like available memory.  So 2.69 GB might not be a correct test.